// pony-router/route/interface.js
module.exports = function(){
    
    this.routes = {};
    this.alternatives = {};
    
    this.route = require('./route');
    
    this.alternative = function(alt, regex){
        this.alternatives[alt] = regex;
    };
    
    this.get = function(route, callback){
        this.routes[route+'%*&get'] = (new this.route()).callback(callback);
        return this.routes[route+'%*&get'];
    };
    
    this.post = function(route, callback){
        this.routes[route+'%*&post'] = (new this.route()).callback(callback);
        return this.routes[route+'%*&post'];    
    };
    
    this.any = function(route, callback){
        this.routes[route+'%*&any'] = (new this.route()).callback(callback);
        return this.routes[route+'%*&any'];
    };
    
    this.add = function(route, callback, type){
        this.routes[route+'%*&'+type] = (new this.route()).callback(callback);
        return this.routes[route+'%*&'+type];
    };
    
    this.regex = function(path){
        
    };
    
    this.generate = function(){
        
        var routes = [];
        
        for(route in this.routes){
            routes.push({
               "path": route.split('%*&')[0],
               "type": route.split('%*&')[1],
               "callback": this.routes[route].callback,
               "before": this.routes[route].before,
               "after": this.routes[route].after
            });
        }
        
        return {
            "routes": routes,
            "alternatives": this.alternatives
        };
    };
}
